docs/Training-Generalization-Learning.md

```diff
@@ -0,0 +1,105 @@
+
+    * `sampler-type` - Specify the sampler type to use for the reset parameter. This is a string that should exist in the `Sampler Factory` (explained below).
+
+    * `sub-arguments` - Specify the characteristic parameters for the sampler. In the example sampler file above, this would correspond to the `intervals` key under the `multirange_uniform` sampler for the gravity reset parameter. The key name should match the name of the corresponding argument in the sampler definition. (Look at defining a new sampler method)
+
+The sampler manager allocates a sampler for a reset parameter by using the *Sampler Factory*, which maintains a dictionary mapping of string keys to sampler objects. The available samplers to be used for reset parameter resampling is as available in the Sampler Factory.
+
+The implementation of the samplers can be found at `ml-agents-envs/mlagents/envs/sampler_class.py`.
+
+### Defining a new sampler method
+
+Custom sampling techniques must inherit from the *Sampler* base class (included in the `sampler_class` file) and preserve the interface. Once the class for the required method is specified, it must be registered in the Sampler Factory. 
+
+This can be done by subscribing to the *register_sampler* method of the SamplerFactory. The command is as follows:
+
+`SamplerFactory.register_sampler(*custom_sampler_string_key*, *custom_sampler_object*)`
+
+Once the Sampler Factory reflects the new register, the custom sampler can be used for resampling reset parameter. For demonstration, lets say our sampler was implemented as below, and we register the `CustomSampler` class with the string `custom-sampler` in the Sampler Factory.
+
+```python
+class CustomSampler(Sampler):
+
+    def __init__(self, argA, argB, argC):
+        self.possible_vals = [argA, argB, argC]
+
+    def sample_all(self):
+        return np.random.choice(self.possible_vals)
+```
+
+Now we need to specify this sampler in the sampler file. Lets say we wish to use this sampler for the reset parameter *mass*; the sampler file would specify the same for mass as the following.
+
+```yaml
+mass:
+    sampler-type: "custom-sampler"
+    argA: 1
+    argB: 2
+    argC: 3
+```
+
+With the sampler file setup, we can proceed to train our agent as explained in the next section.
+
+### Training with Generalization Learning
+
+We first begin with setting up the sampler file. After the sampler file is defined and configured, we proceed by launching `mlagents-learn` and specify our configured sampler file with the `--sampler` flag. To demonstrate, if we wanted to train a 3D ball agent with generalization using the `generalization-test.yaml` sampling setup, we can run
+
+```sh
+mlagents-learn config/trainer_config.yaml --sampler=config/generalize_test.yaml --run-id=3D-Ball-generalization --train
```

Comment:
`generalization-test.yaml` =/= `config/generalize_test.yaml`

Reply:
In the sample code you use config/generalize_test.yaml but in the text you use config/generalization-test.yaml